### PR TITLE
Host ha and multiple management server internal load balancing

### PR DESCRIPTION
**Pull request by Spaceman1984**
Added some documentation around Host HA and management-server internal load balancing

### CODE REVIEW

File: source/adminguide/reliability.rst [hunk 1]
```diff
@@ -78,25 +156,8 @@ HA features work with iSCSI or NFS primary storage. HA with local
 storage is not supported.
 
 
-HA for Hosts
-------------
-
-The user can specify a virtual machine as HA-enabled. By default, all
```
Comment (sureshanaparti) [inline source/adminguide/reliability.rst:84]: @Spaceman1984 any particular reason to remove this text here? Is this doc moved elsewhere?

File: source/adminguide/reliability.rst [hunk 2]
```diff
@@ -61,6 +61,84 @@ still available but the system VMs will not be able to contact the
...
+
+In a Cloudstack environment with multiple management servers, an agent can be
+configured, based on an algorithm, to which management server to connect to.
+This can be useful as an internal loadbalancer or for high availability.
+An administrator is responsible for setting the list of management servers and
+choosing a sorting algorithm using global settings.
+The management server is responsible for propagating the settings to the
+connected agents.
+
+Examples of an agent includes, the process responsible for communication to the
```
Comment (sureshanaparti) [inline source/adminguide/reliability.rst:75]: do you mean by process here?

File: source/adminguide/reliability.rst [hunk 3]
```diff
@@ -61,6 +61,84 @@ still available but the system VMs will not be able to contact the
...
+- After an administrator changes the 'host' and/or the
+'indirect.agent.lb.algorithm' global settings.
+
+On the agent side, the 'host' setting is saved in its properties file as:
+`host=<comma separated addresses>@<algorithm name>`.
+
+From the agent's perspective, the first address in the propagated list
+will be considered the preferred host. A new background task can be
+activated by configuring the `indirect.agent.lb.check.interval` which is
+a cluster level global setting from CloudStack and adminitrators can also
```
Comment (sureshanaparti) [inline source/adminguide/reliability.rst:117]:
```suggestion
a cluster level global setting from CloudStack and administrators can also
```

File: source/adminguide/reliability.rst [hunk 4]
```diff
@@ -61,6 +61,84 @@ still available but the system VMs will not be able to contact the
...
+override this by configuring the 'host.lb.check.interval' in the
+`agent.properties` file.
+
+When an agent gets a host and algorithm combination, the host specific
+background check interval is also sent and is dynamically reconfigured
+in the background task without need to restart agents.
+
+Note: The 'static' and 'roundrobin' algorithms, strictly checks for the
```
Comment (sureshanaparti) [inline source/adminguide/reliability.rst:125]: can you move this algorithm details to top, where they are described ?

File: source/adminguide/reliability.rst [hunk 5]
```diff
@@ -126,6 +187,150 @@ that you want to dedicate to HA-enabled VMs.
...
+any failures. Specific thersholds can be set for failed investigations,
+only when it’s exceeded, will the host transition to a different state.
+
+Host-HA uses both health checks and activity checks to make decisions on 
+recovering and fencing actions. Once determined that the host is in faulty 
+state (health checks failed) it runs activity checks to figure out if there is 
+any disk activity on the VMs running on the specific host.
+
+HA Resource Management Service
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (sureshanaparti) [inline source/adminguide/reliability.rst:220]: resource management service is part of Host HA, I think, no need of a separate section for this.

File: source/adminguide/reliability.rst (hunk 1 above)
Comment (Spaceman1984) [reply source/adminguide/reliability.rst:84]: The title says HA for Hosts, but the content that followed was a duplication of the previous section.

File: source/adminguide/reliability.rst (hunk 2 above)
Comment (Spaceman1984) [reply source/adminguide/reliability.rst:75]: I am trying to describe what I mean by the word "agent". Cloudstack has a few things that are called as such.

File: source/adminguide/reliability.rst (hunk 5 above)
Comment (Spaceman1984) [reply source/adminguide/reliability.rst:220]: ok, I'll remove the heading.

File: source/adminguide/reliability.rst (hunk 4 above)
Comment (Spaceman1984) [reply source/adminguide/reliability.rst:125]: ok

File: source/adminguide/accounts.rst [hunk 6]
```diff
@@ -337,15 +337,15 @@ the user are used.
        #. CloudStack searches for it in LDAP by the configured
           ``ldap.username.attribute``.
```
Comment (sureshanaparti) [inline source/adminguide/accounts.rst:339]: @Spaceman1984 The changes in this file, are made in other PR as well: https://github.com/apache/cloudstack-documentation/pull/225, either remove it from one PR (#223 or #225) or make sure conflicts are resolved properly.

File: source/adminguide/reliability.rst (hunk 2 above)
Comment (sureshanaparti) [reply source/adminguide/reliability.rst:75]: > I am trying to describe what I mean by the word "agent". Cloudstack has a few things that are called as such.

I think, it is better to be specific about the agents.

File: source/adminguide/reliability.rst [hunk 7]
```diff
@@ -61,25 +61,86 @@ still available but the system VMs will not be able to contact the
...
+  production).
+
+.. note:: 
+   The 'static' and 'roundrobin' algorithms, strictly checks for the order as
+   expected by them, however, the 'shuffle' algorithm just checks for content
+   and not the order of the comma separate management server host addresses.
+
+Any changes to the global settings - `indirect.agent.lb.algorithm` and
+`host` does not require restarting of the management server(s) and the
+agents. A change in these global settings will be propagated to all connected
```
Comment (sureshanaparti) [inline source/adminguide/reliability.rst:106]: this seems repeated, it is already mentioned above.

File: source/adminguide/accounts.rst (hunk 6 above)
Comment (Spaceman1984) [reply source/adminguide/accounts.rst:339]: This will not cause a conflict @sureshanaparti

Comment (Spaceman1984) [reply source/adminguide/accounts.rst:339]: I'll remove. no problem

File: source/adminguide/reliability.rst (hunk 2 above)
Comment (Spaceman1984) [reply source/adminguide/reliability.rst:75]: I've added the process name in brackets, does this make more sense?

File: source/adminguide/reliability.rst (hunk 7 above)
Comment (Spaceman1984) [reply source/adminguide/reliability.rst:106]: In this part I'm talking about not having to restart the management server after changing the settings, I didn't mention this before @sureshanaparti .

**Review by sureshanaparti — APPROVED** [review]
changes LGTM

File: source/adminguide/reliability.rst [hunk 8]
```diff
@@ -126,6 +183,107 @@ that you want to dedicate to HA-enabled VMs.
...
+
+Host-HA has granular configuration on a host/cluster/zone level. In a large 
+environment, some hosts from a cluster can be HA-enabled and some not, 
+
+Host-HA uses a state machine design to manage the operations of recovering
+and fencing hosts. The current status of a host is reported when quering a 
+specific host.
+
+Timely health investigations are done on HA-Enabled hosts to monitor for
+any failures. Specific thersholds can be set for failed investigations,
```
Comment (shwstppr) [inline source/adminguide/reliability.rst:207]:
```suggestion
any failures. Specific thresholds can be set for failed investigations,
```

**Review by shwstppr — CHANGES_REQUESTED** [review]
Minor nit. Rest LGTM.
Will be good to have review from @andrijapanicsb

**Review by shwstppr — APPROVED** [review]

**Comment by andrijapanicsb** [comment]
Will do after my vacation pls, yes. On my to-do list (among other doc update reviews)

**Comment by rhtyd** [comment]
ping @andrijapanicsb

**Comment by andrijapanicsb** [comment]
@blueorangutan docbuild

**Review by andrijapanicsb — APPROVED** [review]
LGTM (even though most of the HA-Enabled Hosts section is a copy-paste from the CWIKI design document)

**Pull request merged**